components/overflow-group/overflow-group-mixin.js

```diff
@@ -0,0 +1,349 @@
+export const OVERFLOW_MINI_DROPDOWN_CLASS = 'd2l-overflow-dropdown-mini';
+
+const AUTO_SHOW_CLASS = 'd2l-button-group-show';
+const AUTO_NO_SHOW_CLASS = 'd2l-button-group-no-show';
+
+const OPENER_TYPE = {
+	DEFAULT: 'default',
+	ICON: 'icon'
+};
+
+async function filterAsync(arr, callback) {
+	const fail = Symbol();
+	const results = await Promise.all(arr.map(async item => {
+		const callbackResult = await callback(item);
+		return callbackResult ? item : fail;
+	}));
+	return results.filter(i => i !== fail);
+}
+
+export const OverflowGroupMixin = superclass => class extends LocalizeCoreElement(superclass) {
+
+	static get properties() {
+		return {
+			/**
+			 * Use predefined classes on slot elements to set min and max buttons to show
+			 * @type {boolean}
+			 */
+			autoShow: {
+				type: Boolean,
+				attribute: 'auto-show',
+			},
+			/**
+			 * @ignore
+			 */
+			chompIndex: {
+				type: Number,
+			},
+			/**
+			 * minimum amount of buttons to show
+			 * @type {number}
+			 */
+			minToShow: {
+				type: Number,
+				reflect: true,
+				attribute: 'min-to-show',
+			},
+			/**
+			 * maximum amount of buttons to show
+			 * @type {number}
+			 */
+			maxToShow: {
+				type: Number,
+				reflect: true,
+				attribute: 'max-to-show',
+			},
+			/**
+			 * @ignore
+			 */
+			mini: {
+				type: Boolean,
+				reflect: true
+			},
+			/**
+			 * Set the opener type to 'icon' for a `...` menu icon instead of `More actions` text
+			 * @type {'default'|'icon'}
+			 */
+			openerType: {
+				type: String,
+				attribute: 'opener-type'
+			}
+		};
+	}
+
+	static get styles() {
+		return [offscreenStyles, css`
+			:host {
+				display: block;
+			}
+			:host([hidden]) {
+				display: none;
+			}
+			.d2l-overflow-group-container {
+				display: flex;
+				flex-wrap: wrap;
+				justify-content: var(--d2l-overflow-group-justify-content, normal);
+			}
+			.d2l-overflow-group-container ::slotted([data-is-chomped]) {
+				display: none !important;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+		this._handleItemMutation = this._handleItemMutation.bind(this);
+		this._handleResize = this._handleResize.bind(this);
+
+		this._throttledResize = (entries) => requestAnimationFrame(() => this._handleResize(entries));
+
+		this._overflowHidden = false;
+		this.autoShow = false;
+		this.maxToShow = -1;
+		this.minToShow = 1;
+		this.mini = this.openerType === OPENER_TYPE.ICON;
+		this.openerType = OPENER_TYPE.DEFAULT;
+		this._resizeObserver = null;
+		this._slotItems = [];
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		if (this._resizeObserver) this._resizeObserver.disconnect();
+	}
+
+	async firstUpdated() {
+		super.firstUpdated();
+
+		// selected elements
+		this._buttonSlot = this.shadowRoot.querySelector('slot');
+
+		this._container = this.shadowRoot.querySelector('.d2l-overflow-group-container');
+
+		this._resizeObserver = new ResizeObserver(this._throttledResize);
+		this._resizeObserver.observe(this._container);
```

Comment:
I think we normally wire these up in `connectedCallback`, since `firstUpdated` won't get called again when something is removed and then re-added to the DOM.

Reply:
I think we'll just need to be careful of the timing, since `this._container` until the first render has completed. It would be cleaner if we could wire up the observer to `this`... not sure if that's possible. That element might have been chosen specifically for the observer since it's the one that specifies some layout.

Reply:
Ah yup `this._container` doesn't exist in `connectedCallback`, though if we add in a `requestAnimationFrame` to `connectedCallback` (we do this in `hierarchical-view-mixin`) it does exist so we could do that.

I would be a bit afraid of using `this` instead of `this._container` - I've run into a lot of troubles with similar functionality in `tag-list` figuring out what to put the `ResizeObserver` on and I'd worry that there'd be some weird behaviour that would pop up.